src/spec-configuration/containerFeaturesOrder.ts

```diff
@@ -31,7 +31,7 @@ export function computeOverrideInstallationOrder(config: DevContainerConfig, fea
     // Moves to the beginning the features that are explicitly configured.
     const orderedFeatures = [];
     for (const featureId of config.overrideFeatureInstallOrder!) {
-        const feature = automaticOrder.find(feature => feature.features[0].name === featureId);
+        const feature = automaticOrder.find(feature => feature.features[0].name === featureId || (feature.sourceInformation.type === 'oci' && feature.sourceInformation.featureRef.resource === featureId));
```

Comment:
Would this have to possibility to find the wrong feature? Suppose I have `ghcr.io/devcontainers/features/node` and `ghcr.io/myorg/myrepo/node`. If it's searching by `name` alone (which is `node` in this case?), then it's ambiguous which one it will find.

I think we always want to go by the fully-qualified id and not rely on shorter name matching.

Reply:
You could also add a test for this case with two features uploaded to different registries with the same name/suffix and asserting that you can pin the order correctly for either one.

Reply:
For OCI, the feature would not be resolved with `name`, but instead it will use the newly added condition which checks for `resource`.  (feature.features[0].name="node" and featureId="ghcr.io/devcontainers/features/node or ghcr.io/myorg/myrepo/node"  )

`name` is useful only when referencing a shorthand syntax or local-feature.

![image](https://user-images.githubusercontent.com/24955913/183489585-54fd1020-a97a-4715-b945-e360f3b8f823.png)

Reply:
@samruddhikhandale when using OCI-published features, is the `name` undefined at runtime or would it be `node` in the example above? If it's `node`, then having the `name` check as the first condition will cause the shorter one to short-circuit the fully-qualified name check. That's why I think we'd always want to prefer the fully qualified name with higher precedence, or better yet, only ever match on the fully-qualified name.

Reply:
Ah, sorry! I misunderstood your question. I have updated the condition and added tests, for OCI features it will never resolve with `name` but use `resource` instead. 

![image](https://user-images.githubusercontent.com/24955913/183504092-bc267dbd-aff7-443c-a38e-0b1cea8bf497.png)

Reply:
I think we want to standardize on using `feature[0].id` here, which should always match the `id` of the individual feature's `devcontainer-feature.json`.  The issues I discovered recently was due to that not being consistent in the new code.  For information on _where_ the feature came from, we have the `sourceInformation` object attached to each feature.

I believe we have no standardized pattern for what 'name' should look like.  I have always thought of it as the human-readable name and am surprised we're using it here

Reply:
I am working on additional changes that will always assert that the `feature[0].id` has no special characters https://github.com/devcontainers/cli/blob/jospicer/feature-test-invariants/src/test/container-features/faetureHelpers.test.ts#L39-L44

Reply:
> I think we want to standardize on using `feature[0].id` here, which should always match the `id` of the individual feature's `devcontainer-feature.json`.

@joshspicer Isn't "id" in the devcontainer-feature.json a short name that is ambiguous?

Reply:
Perhaps this need clarification in the spec, but the way `devcontainers/features` is going to be structured, each folder name, `devcontainer-feature.json` id, and the identifier that would go in a user's devcontainer should all be the same and unique within the collection.  We will parse out `sourceInformation` based on how the user references the feature in their devcontainer.json

So for example,

`ghcr.io/devcontainer/features/ruby:1`

the `id` of this feature MUST be `ruby`, and its source code MUST be enclosed in a folder named `ruby`.

(https://github.com/devcontainers/features/tree/main/src/ruby)

Reply:
> I think we want to standardize on using feature[0].id here, which should always match the id of the individual feature's devcontainer-feature.json

@joshspicer According to the [spec](https://github.com/devcontainers/spec/blob/main/proposals/devcontainer-features.md#referencing-a-feature), the `overrideFeatureInstallOrder` expects list of feature IDs. In case of a local feature (aka file-path), it expects `<path-to-feature-dir>` which can only be resolved by matching the `feature[0].name`
  
![image](https://user-images.githubusercontent.com/24955913/183703976-1f7fb014-4710-4f2c-b8df-78f8c377c7e8.png)

Similarly, for tgz, the automatic order can be resolved only on `feature[0].name` - [Reference](https://github.com/devcontainers/cli/blob/main/src/spec-configuration/containerFeaturesConfiguration.ts#L571)

Reply:
Looking at [processFeatureIdentifier()](https://github.com/devcontainers/cli/blob/main/src/spec-configuration/containerFeaturesConfiguration.ts#L520), the `feature.name` is not necessarily the `name` param from `devcontainer.json` 🤔

Reply:
My view is that:

- `name` should never be used for anything other than UX/human-readable purposes.  Any current uses of it in the CLI for processing is invalid and should be fixed.
- `id` is the unique name of a feature in a given collection. _It contains NO information about _where_ this feature came from_.
- Each feature object also has a `sourceInformation` object attached to it, which is the result of parsing the user's devcontainer to separate _where_ the feature came from, and the feature itself.

Reply:
A `file-path` feature has a `resolveFilePath` attribute that points to the feature's location on disk.  https://github.com/devcontainers/cli/blob/main/src/spec-configuration/containerFeaturesConfiguration.ts#L93-L96

Be sure not to get this with the antiquate 'local-cache' `SourceInformation` type.  Those are the packaged features that we ship directly with the CLI (https://github.com/devcontainers/cli/blob/main/src/spec-configuration/containerFeaturesConfiguration.ts#L78-L80)

Reply:
> Isn't "id" in the devcontainer-feature.json a short name that is ambiguous?

@chrmarti It _would_ be ambiguous, but every Feature has a `SourceInformation` object attached to it that indicates where it was downloaded/sourced from.  

@edgonmsft 's solution to ensuring the node layer can handle >1 feature unambiguously is to name each temporary build folder uniquely
https://github.com/devcontainers/cli/blob/main/src/spec-configuration/containerFeaturesConfiguration.ts#L715-L722

Reply:
> A file-path feature has a resolveFilePath attribute that points to the feature's location on disk.

For local feature scenario, the `resolveFilePath`  is not necessarily same as the id provided in `overrideFeatureInstallOrder`

![image](https://user-images.githubusercontent.com/24955913/183753057-ba8929ac-3e5b-4c65-bc97-ddf7ecdd736f.png)

We can definitely not use `id` because how can we differentiate `./local-feature/node` with `/local/node`

Reply:
Discussed offline with @joshspicer , I'll add another property `userFeatureId` in `SourceInformation` which will provide us with the `id` expected by https://github.com/devcontainers/spec/blob/main/proposals/devcontainer-features.md#referencing-a-feature. This prop will be used in the above condition (no need of `feature[0].name`).

I'll create a followup PR to modify the `processFeatureIdentifier` function which should take in a DevContainerFeature (the raw JSON from the user's devcontainer), and return a feature set with all the relevant information parsed out.